agent-shell.el

```diff
@@ -2029,6 +2039,71 @@ When DEACTIVATE is non-nil, deactivate region/selection."
+
+(defun agent-shell--get-mode-name (mode-id available-modes)
+  "Get the name of the mode with MODE-ID from AVAILABLE-MODES.
+AVAILABLE-MODES is the list of mode objects from the ACP session/new response.
+Each mode has an 'id' and 'name' field. We look up the mode by ID to get its
+display name."
+  (when-let ((mode (cl-find mode-id available-modes
+                            :key (lambda (m) (map-elt m 'id))
+                            :test #'string=)))
+    (map-elt mode 'name)))
+
+(defun agent-shell--mode-line-format ()
+  "Return the mode-line format for displaying the current permission mode."
+  (when (derived-mode-p 'agent-shell-mode)
+    (let* ((state (agent-shell--state))
+           (current-mode-id (map-elt state :current-mode-id))
+           (available-modes (map-elt state :available-modes)))
+      (when-let ((mode-name (and current-mode-id available-modes
+                                 (agent-shell--get-mode-name current-mode-id available-modes))))
+        (propertize (format " [%s]" mode-name)
+                    'face 'font-lock-constant-face
+                    'help-echo (format "ACP Permission Mode: %s" mode-name))))))
+
+(defun agent-shell--setup-modeline ()
+  "Set up the modeline to display permission mode.
+Uses :eval so the mode updates automatically when state changes."
+  (setq-local mode-line-misc-info
+              (append mode-line-misc-info
+                      '((:eval (agent-shell--mode-line-format))))))
+
+(defun agent-shell-cycle-mode ()
+  "Cycle through available permission modes for the current agent-shell session."
+  (interactive)
+  (unless (derived-mode-p 'agent-shell-mode)
+    (user-error "Not in an agent-shell buffer"))
+  (let* ((state (agent-shell--state))
+         (session-id (map-elt state :session-id))
+         (available-modes (map-elt state :available-modes))
+         (current-mode-id (map-elt state :current-mode-id)))
+    (unless session-id
+      (user-error "No active session"))
+    (unless available-modes
+      (user-error "No modes available"))
+    ;; Find next mode
+    (let* ((mode-ids (mapcar (lambda (mode) (map-elt mode 'id)) available-modes))
+           (current-idx (or (cl-position current-mode-id mode-ids :test #'string=) -1))
+           (next-idx (mod (1+ current-idx) (length mode-ids)))
+           (next-mode-id (nth next-idx mode-ids))
+           (next-mode (cl-find next-mode-id available-modes
+                               :key (lambda (mode) (map-elt mode 'id))
+                               :test #'string=)))
+      ;; Send request to change mode
+      (acp-send-request
+       :client (map-elt state :client)
+       :request (acp-make-session-set-mode-request
+                 :session-id session-id
+                 :mode-id next-mode-id)
+       :on-success (lambda (_response)
+                     (setf (map-elt agent-shell--state :current-mode-id) next-mode-id)
+                     (with-current-buffer (map-elt state :buffer)
```

Comment:
We can omit with-current-buffer here, right? Lemme know if otherwise as it would be a bug.